src/MongoDB.Driver/Linq/Linq3Implementation/Ast/PipelineOptimizer/AstGroupPipelineOptimizer.cs

```diff
@@ -0,0 +1,413 @@
+﻿/* Copyright 2010-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using MongoDB.Bson;
+using MongoDB.Driver.Linq.Linq3Implementation.Ast.Expressions;
+using MongoDB.Driver.Linq.Linq3Implementation.Ast.Filters;
+using MongoDB.Driver.Linq.Linq3Implementation.Ast.Stages;
+using MongoDB.Driver.Linq.Linq3Implementation.Ast.Visitors;
+
+namespace MongoDB.Driver.Linq.Linq3Implementation.Ast.PipelineOptimizer
+{
+    internal class AstGroupPipelineOptimizer
+    {
+        private readonly AccumulatorSet _accumulators = new AccumulatorSet();
```

Comment:
The new `AccumulatorSet` helper class helps hide a small amount of abstraction around building up a set of accumulators (including hiding the trick of using `Render` to compare two accumulator expressions).